## Introduction to DataFrames

In [1]:
import pandas as pd

In [2]:
nba = pd.read_csv("nba.csv")

In [3]:
nba.head(10)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

             College      Salary  
0              Texas   7730337.0  
1          Marquette   6796117.0  
2  Boston University         NaN  
3      Georgia State   1148640.0  
4                NaN   5000000.0  
5                NaN  12000000.0  
6                LSU   1170960.0  
7            Gonzaga   2165160.0  
8         Louisville   1824360.0  
9     Oklahoma State   3431040.0

In [4]:
nba.values

array([['Avery Bradley', 'Boston Celtics', 0.0, ..., 180.0, 'Texas',
        7730337.0],
       ['Jae Crowder', 'Boston Celtics', 99.0, ..., 235.0, 'Marquette',
        6796117.0],
       ['John Holland', 'Boston Celtics', 30.0, ..., 205.0,
        'Boston University', nan],
       ...,
       ['Tibor Pleiss', 'Utah Jazz', 21.0, ..., 256.0, nan, 2900000.0],
       ['Jeff Withey', 'Utah Jazz', 24.0, ..., 231.0, 'Kansas', 947276.0],
       [nan, nan, nan, ..., nan, nan, nan]], dtype=object)

In [5]:
nba.shape

(458, 9)

this is a 2-D array of data set

In [6]:
nba.dtypes #to get data type info of the fields

Name         object
Team         object
Number      float64
Position     object
Age         float64
Height       object
Weight      float64
College      object
Salary      float64
dtype: object

In [7]:
nba.columns

Index(['Name', 'Team', 'Number', 'Position', 'Age', 'Height', 'Weight',
       'College', 'Salary'],
      dtype='object')

this gives you all the column(fields) names

In [8]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null object
Number      457 non-null float64
Position    457 non-null object
Age         457 non-null float64
Height      457 non-null object
Weight      457 non-null float64
College     373 non-null object
Salary      446 non-null float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


##### this is like the describe() method in the series object

In [12]:
rev = pd.read_csv("revenue.csv", index_col= "Date")

In [13]:
rev.head()

New York  Los Angeles  Miami
Date                                
1/1/16       985          122    499
1/2/16       738          788    534
1/3/16        14           20    933
1/4/16       730          904    885
1/5/16       114           71    253

To get the total sum of rev for each state

In [14]:
rev.sum()

New York       5475
Los Angeles    5134
Miami          5641
dtype: int64

this sums on the y-axis by default. However, if we want to sum for all three states in a row (i.e x-axis)

In [15]:
rev.sum(axis = 1)

Date
1/1/16     1606
1/2/16     2060
1/3/16      967
1/4/16     2519
1/5/16      438
1/6/16     1935
1/7/16     1234
1/8/16     2313
1/9/16     2623
1/10/16     555
dtype: int64

the axis parameter determines how it sums the values

### Extracting data from a field (or fields) in DataFrames

In [16]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [18]:
nba["Number"].head()

0     0.0
1    99.0
2    30.0
3    28.0
4     8.0
Name: Number, dtype: float64

notice: since you selected just one field, it returned a series (1-D)

In [19]:
nba[["Number", "Position", "Age"]].tail(10)

Number Position   Age
448    20.0       SF  26.0
449     5.0       SG  23.0
450     2.0       SF  28.0
451    23.0       SF  26.0
452    41.0       PF  20.0
453     8.0       PG  26.0
454    25.0       PG  24.0
455    21.0        C  26.0
456    24.0        C  26.0
457     NaN      NaN   NaN

see how we selected multiple fields

In [29]:
nba[nba["Number"] > 30][["Name","Number","Position","Age"]].head(10)

Name  Number Position   Age
1              Jae Crowder    99.0       SF  25.0
5             Amir Johnson    90.0       PF  29.0
6            Jordan Mickey    55.0       PF  21.0
7             Kelly Olynyk    41.0        C  25.0
9             Marcus Smart    36.0       PG  22.0
14            Tyler Zeller    44.0        C  26.0
15        Bojan Bogdanovic    44.0       SG  27.0
25             Willie Reed    33.0       PF  26.0
26         Thomas Robinson    41.0       PF  25.0
32  Thanasis Antetokounmpo    43.0       SF  23.0

the code above shows how to filter data from the Dataframe

In [30]:
nba.columns

Index(['Name', 'Team', 'Number', 'Position', 'Age', 'Height', 'Weight',
       'College', 'Salary'],
      dtype='object')

In [31]:
nba[nba["Number"] > 30].head()

Name            Team  Number Position   Age Height  Weight  \
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
5   Amir Johnson  Boston Celtics    90.0       PF  29.0    6-9   240.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

          College      Salary  
1       Marquette   6796117.0  
5             NaN  12000000.0  
6             LSU   1170960.0  
7         Gonzaga   2165160.0  
9  Oklahoma State   3431040.0

### Add New Columns to a DataFrame

In [32]:
nba["Sport"] = "Basketball"

In [33]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary       Sport  
0              Texas  7730337.0  Basketball  
1          Marquette  6796117.0  Basketball  
2  Boston University        NaN  Basketball  
3      Georgia State  1148640.0  Basketball  
4                NaN  5000000.0  Basketball

In [34]:
nba["Age_Double"] = nba["Age"] * 2

In [35]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary       Sport  Age_Double  
0              Texas  7730337.0  Basketball        50.0  
1          Marquette  6796117.0  Basketball        50.0  
2  Boston University        NaN  Basketball        54.0  
3      Georgia State  1148640.0  Basketball        44.0  
4                NaN  5000000.0  Basketball        58.0

In [36]:
del nba["Sport"]

In [37]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  Age_Double  
0              Texas  7730337.0        50.0  
1          Marquette  6796117.0        50.0  
2  Boston University        NaN        54.0  
3      Georgia State  1148640.0        44.0  
4                NaN  5000000.0        58.0

##### You can also use the .insert() method

nba.insert(loc, column, value) 

where:<br> 
>loc = index position you want to place field<br>
>column = name of column<br>
>value = data

In [39]:
nba.insert(7, column = "Weight(Kg)",value = nba["Weight"] * 0.453592)

In [40]:
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

   Weight(Kg)            College     Salary  Age_Double  
0   81.646560              Texas  7730337.0        50.0  
1  106.594120          Marquette  6796117.0        50.0  
2   92.986360  Boston University        NaN        54.0  
3   83.914520      Georgia State  1148640.0        44.0  
4  104.779752                NaN  5000000.0        58.0

##### Using .value_counts() method in DataFrames

In [46]:
nba["Team"].value_counts().head(7).sort_values()

Charlotte Hornets       15
Boston Celtics          15
San Antonio Spurs       15
New York Knicks         16
Milwaukee Bucks         16
Memphis Grizzlies       18
New Orleans Pelicans    19
Name: Team, dtype: int64

In [47]:
nba[nba["Team"] == "Boston Celtics"]["Team"].value_counts()

Boston Celtics    15
Name: Team, dtype: int64

#### Drop Rows with Null Values

Using the **.dropna()** method to drop rows with NaN values

In [48]:
nba.tail()

Name       Team  Number Position   Age Height  Weight  \
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0   
457           NaN        NaN     NaN      NaN   NaN    NaN     NaN   

     Weight(Kg) College     Salary  Age_Double  
453   92.079176  Butler  2433333.0        52.0  
454   81.192968     NaN   900000.0        48.0  
455  116.119552     NaN  2900000.0        52.0  
456  104.779752  Kansas   947276.0        52.0  
457         NaN     NaN        NaN         NaN

In [50]:
nba.dropna().tail()

Name       Team  Number Position   Age Height  Weight  \
449    Rodney Hood  Utah Jazz     5.0       SG  23.0    6-8   206.0   
451  Chris Johnson  Utah Jazz    23.0       SF  26.0    6-6   206.0   
452     Trey Lyles  Utah Jazz    41.0       PF  20.0   6-10   234.0   
453   Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0   
456    Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0   

     Weight(Kg)   College     Salary  Age_Double  
449   93.439952      Duke  1348440.0        46.0  
451   93.439952    Dayton   981348.0        52.0  
452  106.140528  Kentucky  2239800.0        40.0  
453   92.079176    Butler  2433333.0        52.0  
456  104.779752    Kansas   947276.0        52.0


Notice how row 457 is dropped because it has NaN values

In [52]:
nba.dropna(how = "all").head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

   Weight(Kg)            College     Salary  Age_Double  
0   81.646560              Texas  7730337.0        50.0  
1  106.594120          Marquette  6796117.0        50.0  
2   92.986360  Boston University        NaN        54.0  
3   83.914520      Georgia State  1148640.0        44.0  
4  104.779752                NaN  5000000.0        58.0

The .dropna() method removes any row with one or more NaN values by default. However, the __how__ parameter can be used to tweak it. As seen above. It removes only rows that all its fields have NaN values.<br> It also takes in the __inplace__ parameter to make changes permanent.

In [62]:
nba.dropna(axis = 1).head() #this will remove fields(or columns) with NaN values

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4]

To remove rows that have NaN in particular fields, we use the __subset__ param

In [70]:
nba = pd.read_csv("nba.csv")

In [77]:
nba[nba["Salary"] == 7730337.0]

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   

  College     Salary  
0   Texas  7730337.0

In [96]:
nba[nba["College"].isnull()].sort_values(by ='Name').head()

Name                   Team  Number Position   Age Height  \
330       Al Jefferson      Charlotte Hornets    25.0        C  31.0   6-10   
275      Alexis Ajinca   New Orleans Pelicans    42.0        C  28.0    7-2   
348  Amar'e Stoudemire             Miami Heat     5.0       PF  33.0   6-10   
5         Amir Johnson         Boston Celtics    90.0       PF  29.0    6-9   
90    Anderson Varejao  Golden State Warriors    18.0       PF  33.0   6-11   

     Weight College      Salary  
330   289.0     NaN  13500000.0  
275   248.0     NaN   4389607.0  
348   245.0     NaN    947276.0  
5     240.0     NaN  12000000.0  
90    273.0     NaN    289755.0

In [98]:
nba.dropna(subset= ["College","Salary"]).head(7)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
8   Terry Rozier  Boston Celtics    12.0       PG  22.0    6-2   190.0   
9   Marcus Smart  Boston Celtics    36.0       PG  22.0    6-4   220.0   

          College     Salary  
0           Texas  7730337.0  
1       Marquette  6796117.0  
3   Georgia State  1148640.0  
6             LSU  1170960.0  
7         Gonzaga  2165160.0  
8      Louisville  1824360.0  
9  Oklahoma State  3431040.0

The above command drops all rows with NaN under __College__ and __Salary__ fields

##### Fill in Null values with the __.fillna()__ method<br> If you have a consistent dataType for all fields, you can use this method in the DF as a whole.<br>If you do not, use it on individual fields

In [116]:
nba["College"].fillna(value="No college", inplace= True)            

The above code fills all the NaN values in the __College__ field with "No college"

#### Using the .astype() method to change dataTypes of fields in a DF

- This method does not act on fields that have NaN values
- First check the DF fields by using .info attribute
- Take off rows with NaN all through with .dropna() method
- Change NaN values in fields using .fillna() method

In [130]:
nba = pd.read_csv("nba.csv")

In [131]:
nba.info() #First check the DF fields by using .info attribute

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null object
Number      457 non-null float64
Position    457 non-null object
Age         457 non-null float64
Height      457 non-null object
Weight      457 non-null float64
College     373 non-null object
Salary      446 non-null float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


In [132]:
nba.dropna(how= "all", inplace= True) #Take off rows with NaN all through with .dropna() method

In [135]:
nba["College"].fillna(value="No College", inplace = True)
nba["Salary"].fillna(value= 0, inplace = True) #Change NaN values in fields using .fillna() method

In [139]:
nba["Salary"].astype("int").head()

0    7730337
1    6796117
2          0
3    1148640
4    5000000
Name: Salary, dtype: int32

Here, we change the dtype of _Salary_ field from float to int. Note, this does not affect the field in the DF.

In [144]:
nba["Salary"] = nba["Salary"].astype("int")
nba["Age"] = nba["Age"].astype("int")
nba["Number"] = nba["Number"].astype("int")

In [145]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
Name        457 non-null object
Team        457 non-null object
Number      457 non-null int32
Position    457 non-null object
Age         457 non-null int32
Height      457 non-null object
Weight      457 non-null float64
College     457 non-null object
Salary      457 non-null int32
dtypes: float64(1), int32(3), object(5)
memory usage: 30.3+ KB


In [149]:
nba["College"].value_counts().head()

No College        84
Kentucky          22
Duke              20
Kansas            18
North Carolina    16
Name: College, dtype: int64

In [150]:
nba["College"].nunique()

119

the __.nunique()__ method returns how many unique data in a field